src/ytypes.i

```diff
@@ -0,0 +1,108 @@
+%{
+YCPValue pyval_to_ycp(PyObject *input)
+{
+    int ret;
+    void *arg = 0;
+
+    if (input == Py_None)
+        return YCPNull();
+    if (PyBool_Check(input)) {
+        if (PyObject_Compare(input, Py_True) == 0)
+            return YCPBoolean(true);
+        else
+            return YCPBoolean(false);
+    }
+    if (PyInt_Check(input))
+        return YCPInteger(PyInt_AsLong(input));
+    if (PyFloat_Check(input))
+        return YCPFloat(PyFloat_AsDouble(input));
+    if (PyString_Check(input))
+        return YCPString(PyString_AsString(input));
+    if (PyList_Check(input)) {
+        auto size = PyList_Size(input);
+        if (size > 0 && PyFunction_Check(PyList_GetItem(input, 0))) {
+            auto t = PyTuple_New(size);
+            for (int i = 0; i < size; i++)
+                PyTuple_SetItem(t, i, PyList_GetItem(input, i));
+            return YCPCode(new YPythonCode(t));
+        } else {
+            YCPList l;
+            for (int i = 0; i < size; i++)
+                l->add(pyval_to_ycp(PyList_GetItem(input, i)));
+            return l;
+        }
+    }
+    if (PyFunction_Check(input))
+        return YCPCode(new YPythonCode(PyTuple_Pack(1, input)));
+    if (PyDict_Check(input)) {
+        YCPMap m;
+        if (PyDict_Size(input) == 0)
+            return m;
+
+        PyObject *key, *value;
+        Py_ssize_t pos = 0;
+        while (PyDict_Next(input, &pos, &key, &value))
+            m->add(pyval_to_ycp(key), pyval_to_ycp(value));
+        return m;
+    }
+    if (PyTuple_Check(input)) {
+        auto size = PyTuple_Size(input);
+        YCPList l;
+        for (int i = 0; i < size; i++)
+            l->add(pyval_to_ycp(PyTuple_GetItem(input, i)));
+        return l;
+    }
```

Comment:
question. I know that you intentionally does not have symbol, but how you pass it to component system if you need to call method like `UI.QueryWidget(:id, :Value)` how you create it in python?

Reply:
How this is handled is a little tricky, but it does work. You would call:
UI.QueryWidget(Id(myid), 'Value')
OR
UI.QueryWidget(Id(myid), Symbol('Value'))

Either way is equivalent. The Symbol class has been wrapped via swig, so we can generated a YCPSymbol manually when necessary (the second example).
The first example works because of some trickery in CallYCPFunction (I'll comment on that line so you can see).

Reply:
ok, so it works.